mock/channels/channelService.js

```diff
@@ -1,40 +1,138 @@
+    await this.closeWindowOnCompletion();
 
-class UserChannelService {
-    async joinChannel() {
-        const channels = await window.fdc3.getSystemChannels();
-        await window.fdc3.joinChannel(channels[0].id);
-        return channels[0];
+    //notify app A that ChannelsApp has finished executing
+    if (config.notifyAppAOnCompletion) {
+      await this.notifyAppAOnCompletion();
     }
+  }
 
-    async broadcast(contextType, historyItem) {
-        await window.fdc3.broadcast({
-            "type": contextType,
-            "name": `History-item-${historyItem}`
-        });
-    }
+  async joinSystemChannelOne() {
+    const channels = await window.fdc3.getSystemChannels();
+    await window.fdc3.joinChannel(channels[0].id);
+    return channels[0];
+  }
+
+  //retrieve/create "test-channel" app channel
+  async retrieveTestAppChannel() {
+    return await window.fdc3.getOrCreateChannel("test-channel");
+  }
+
+  //get broadcast service and broadcast the given context type
+  async broadcastContextItem(contextType, channel, historyItems, testId) {
+    let broadcastService = this.getBroadcastService(channel.type);
+    await broadcastService.broadcast(contextType, historyItems, channel, testId);
+  }
 
-    async addContextListener(contextType) {
-        await window.fdc3.addContextListener(
-            contextType,
-            () => closeFinsembleWindow());
+  //get app/system channel broadcast service
+  getBroadcastService(currentChannelType) {
+    if (currentChannelType === channelType.system) {
+      return this.systemChannelBroadcastService;
+    } else if (currentChannelType === channelType.app) {
+      return this.appChannelBroadcastService;
     }
+  }
+
+  //app channel broadcast service
+  appChannelBroadcastService = {
+    broadcast: async (contextType, historyItems, channel, testId) => {
+      if (channel !== undefined) {
+        for (let i = 0; i < historyItems; i++) {
+          let context = {
+            type: contextType,
+            name: `History-item-${i + 1}`,
+          };
+          if(testId) context = context.testId = testId;
+          await channel.broadcast(context);
+        }
+      }
+    },
+  };
+
+  //system channel broadcast service
+  systemChannelBroadcastService = {
+    broadcast: async (contextType, historyItems, ignored, testId) => {
+      for (let i = 0; i < historyItems; i++) {
+        let context = {
+          type: contextType,
+          name: `History-item-${i + 1}`,
+        };
+        if(testId) context = context.testId = testId;
+        await window.fdc3.broadcast(context);
+      }
+    },
+  };
+
+  //close ChannelsApp on completion and respond to app A
+  async closeWindowOnCompletion(testId) {
+    const appControlChannel = await window.fdc3.getOrCreateChannel(
+      "app-control"
+    );
+    await appControlChannel.addContextListener("closeWindow", async () => {
+      appControlChannel.broadcast({ type: "windowClosed", testId: testId });
```

Comment:
`appControlChannel.broadcast()` can go after `window.close()`. I twill execute even after the window has closed. I tested this out and it worked

Reply:
That may work in one container but not another, it is literally racing against the application's main thread closing the window. Within a window javascript is single threaded, in a multi-window environment different windows may be in different render threads and in an electron app the main process is definitely a different thread. Hence, I think its a bad idea to rely on always winning a race.